app/src/org/commcare/android/view/c3/DataConfiguration.java

```diff
@@ -0,0 +1,349 @@
+            xValues.put(parseXValue(a.getX(), description));
+            yValues.put(parseYValue(a.getY(), description));
+            text.put(a.getAnnotation());
+        }
+
+        mColumns.put(xValues);
+        mColumns.put(yValues);
+        mVariables.put("annotations", text.toString());
+    }
+
+    /**
+     * Create fake series so there's data all the way to the edges of the user-specified
+     * min and max. C3 does tick placement in part based on data, so this will force
+     * it to place ticks based on the user's desired min/max range.
+     * @throws JSONException
+     * @throws InvalidStateException
+     */
+    private void addBoundaries() throws JSONException, InvalidStateException {
+        String xMin = mData.getConfiguration("x-min");
+        String xMax = mData.getConfiguration("x-max");
+
+        // If we don't have user-specified bounds, don't bother.
+        if (xMin == null || xMax == null) {
+            return;
+        }
+
+        String xID = "boundsX";
+        if (addBoundary(xID, "boundsY", "y") || addBoundary(xID, "boundsY2", "secondary-y")) {
+            // If at least one y axis had boundaries and therefore a series was created,
+            // now create the matchin x values
+            JSONArray xValues = new JSONArray();
+            xValues.put(xID);
+            xValues.put(parseXValue(xMin, "x-min"));
+            xValues.put(parseXValue(xMax, "x-max"));
+            mColumns.put(xValues);
+        }
+    }
+
+    /**
+     * Helper for addBoundaries: possibly add a series for either the primary or secondary y axis,
+     * depending on whether or not that axis has a min and max specified.
+     * @param xID ID of x column to associate with the new series
+     * @param yID ID of y column for new series
+     * @param prefix "y" or "secondary-y"
+     * @return True iff a series was actually created
+     * @throws JSONException
+     * @throws InvalidStateException
+     */
+    private boolean addBoundary(String xID, String yID, String prefix) throws JSONException, InvalidStateException {
+        String min = mData.getConfiguration(prefix + "-min");
+        String max = mData.getConfiguration(prefix + "-max");
+        if (min != null && max != null) {
+            mXs.put(yID, xID);
+            mTypes.put(yID, "line");
+            mAxes.put(yID, "y2");
+
+            JSONArray yValues = new JSONArray();
+            yValues.put(yID);
+            yValues.put(parseYValue(min, "secondary-y-min"));
+            yValues.put(parseYValue(max, "secondary-y-max"));
+            mColumns.put(yValues);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Comparator to sort XYPointData-derived objects by x value without parsing them.
+     * Useful for bar graphs, where x values are text.
+     *
+     * @author jschweers
+     */
+    private class StringPointComparator implements Comparator<XYPointData> {

```

Comment:
private classes should be made static whenever possible to prevent the automatic creation of a pointer to the encapsulating class, which sometimes leaks memory